[WIP] The merge interop spec

MarekM25 opened the pull request:
This branch **shouldn't be merged** to the master now. We need to refactor code in several places, and it contains the timestamp hack for the workshop.

Usage:
```
git clone https://github.com/NethermindEth/nethermind --recursive
cd nethermind/src/Nethermind
git checkout themerge
dotnet build Nethermind.sln -c Release
cd Nethermind.Runner
dotnet run -c Release -- --config themerge_devnet
```

For testing purposes we can use test vectors (we use different port 8550 instead of 8545):
https://notes.ethereum.org/@9AeMAlpyQYaAAyuj47BzRw/rkwW3ceVY

**Prepare payload**
`curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_preparePayload","params":[{"parentHash":"0xa0513a503d5bd6e89a144c3268e5b7e9da9dbf63df125a360e3950a7d0d67131", "timestamp":"0x5", "random":"0x0000000000000000000000000000000000000000000000000000000000000000", "feeRecipient":"0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b"}],"id":67}' http://localhost:8550`

**Get payload**
`curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_getPayload","params":["0x0"],"id":67}' http://localhost:8550`

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/BlockConfirmationManager.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        public Keccak LastConfirmedHash { get; private set; } = Keccak.Zero;
+
+        public void Confirm(BlockHeader? blockHeader)
```
take hash as an input maybe? minimum required argument?

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/IBlockConfirmationManager.cs:
```diff
@@ -0,0 +1,39 @@
+        void Confirm(BlockHeader? blockHeader);
+    }
+    
+    public class NoBlockConfirmation : IBlockConfirmationManager
```
BlockConfirmation.None or No.BlockCOnfirmation would feel more fluent

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs:
```diff
@@ -278,20 +278,20 @@ private IEnumerable<Transaction> GetTransactions(BlockHeader parent)
                 parent.Hash!,
                 Keccak.OfAnEmptySequenceRlp,
-                Sealer.Address,
+                blockAuthor ?? Sealer.Address,
```
should there be no PosSealer that would pick the block author properly?

MarekM25 replied:
@kristofgazso

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/Validators/HeaderValidator.cs:
```diff
@@ -107,10 +112,15 @@ public bool Validate(BlockHeader header, BlockHeader? parent, bool isUncle = fal
             // seal is validated when synchronizing so we can remove it from here - review and test
-            bool sealParamsCorrect = _sealValidator.ValidateParams(parent, header);
-            if (!sealParamsCorrect)
+            bool sealParamsCorrect = true;
```
maybe provide a SealValidator.AlwaysValid when in PoS?

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/Validators/HeaderValidator.cs:
```diff
@@ -107,10 +112,15 @@ public bool Validate(BlockHeader header, BlockHeader? parent, bool isUncle = fal
-            bool sealParamsCorrect = _sealValidator.ValidateParams(parent, header);
-            if (!sealParamsCorrect)
+            bool sealParamsCorrect = true;
+            if (_poSSwitcher.IsPos(header) == false)
```
shall we not have a PoS / not PoS on the header itself as a property?

tkstanczak commented on src/Nethermind/Nethermind.Blockchain/Validators/HeaderValidator.cs:
```diff
@@ -221,5 +222,52 @@ private bool ValidateGenesis(BlockHeader header)
+        }
+        private bool ValidateHeaderField<T>(BlockHeader header, T value, T expected, string name)
+        {
+            if (Equals(value, expected)) return true;
```
hmmmm, I wonder about this Equals... it may end up super slow for structs

tkstanczak commented on src/Nethermind/Nethermind.Consensus/ConstantDifficultyCalculator.cs:
```diff
@@ -24,6 +24,8 @@ public class ConstantDifficultyCalculator : IDifficultyCalculator
     {
```
rename to ConstantDifficulty instead of ConstantDifficultyCalculator -> or even -> Difficulty.One, Difficulty.Zero

tkstanczak commented on src/Nethermind/Nethermind.Consensus/IPoSSwitcher.cs:
```diff
@@ -0,0 +1,31 @@
+        bool TrySwitchToPos(BlockHeader header);
+        bool IsPos(BlockHeader header);
+
+        bool WasEverInPoS();
```
HasEverBeenInPos

tkstanczak commented on src/Nethermind/Nethermind.Consensus/NoPoS.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public static NoPoS Instance { get; } = new();
+        
+        public bool TrySwitchToPos(BlockHeader header) => false;
```
should it throw exception or return true?

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Setup.cs:
```diff
@@ -92,6 +100,19 @@ protected override IBlockProducer CreateTestBlockProducer(TxPoolTxSource txPoolT
                     miningConfig,
                     LogManager);
                 
+                EmptyBlockProducer = new Eth2EmptyBlockProducerFactory().Create(
```
also empty block producer, hmmm should it be normal producer with EmptyTxSource?

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Data/ConsensusValidationStatus.cs:
```diff
@@ -15,12 +15,14 @@
-        public bool Valid { get; set; }
-
-        public static implicit operator NewBlockResult(bool value) => new() {Valid = value};
+        Invalid = 0,
```
is 0 invalid again? :)

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutePayloadResult.cs:
```diff
@@ -15,14 +15,20 @@
         public Keccak BlockHash { get; set; }
 
-        public VerificationStatus Status { get; set; } 
+        public string? Status => Enum.GetName(typeof(VerificationStatus), EnumStatus)?.ToUpper();
```
can we make it inmutable and hardcode status instead of allocating strings each time we execute a getter?

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Data/ForkChoiceUpdatedRequest.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            HeadBlockHash = headBlockHash;
+            FinalizedBlockHash = finalizedBlockHash;
+            //ConfirmedBlockHash = confirmedBlockHash;
```
you need the commented out code? add comments why it is needed

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Data/PayloadManager.cs:
```diff
@@ -0,0 +1,106 @@
+        public PayloadManager(IBlockTree blockTree)
+        {
+            _blockTree = blockTree;
+            _executePayloadHashes = new LruCache<Keccak, bool>(300, "Recent Execute Payload Hashes");
```
move 300 to a const and explain why it is 300 and what it means

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Handlers/ConsensusValidatedHandler.cs:
```diff
@@ -0,0 +1,53 @@
+            }
+            else
+            {
+                _payloadManager.TryAddConsensusValidatedResult(request.BlockHash, isValid);
```
what if it fails to add?

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/Handlers/PayloadStorage.cs:
```diff
@@ -0,0 +1,78 @@
+            {
+                while (_payloadStorage.ContainsKey(_currentPayloadId))
+                {
+                    if (_currentPayloadId == uint.MaxValue)
```
inifinite loop possible

tkstanczak commented on src/Nethermind/Nethermind.Merge.Plugin/MergePlugin.BlockProducer.cs:
```diff
@@ -66,6 +69,21 @@ public Task<IBlockProducer> InitBlockProducer(IBlockProductionTrigger? blockProd
                     _api.LogManager
                 );
+                    
+                _emptyBlockProductionTrigger = new BuildBlocksWhenRequested();
```
why not same as default initialization?

tkstanczak commented on src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs:
```diff
@@ -154,13 +170,38 @@ GetBlockHeadersMessage getBlockHeadersMessage
+            }
+            
+            if (_poSSwitcher.WasEverInPoS())
+            {
```
is it ok to enter from two threads here?

marcindsobczak replied:
Shouldn't be a problem, it will then set WasEverInPoS to true twice and will unsubscribe from events in SyncServer twice (so in second try will just do nothing)

MarekM25 commented on src/Nethermind/Nethermind.Blockchain/BlockTree.cs:
```diff
@@ -511,7 +511,7 @@ public void Insert(IEnumerable<Block> blocks)
         }
 
-        private AddBlockResult Suggest(Block? block, BlockHeader header, bool shouldProcess = true, bool? setAsMain = null)
+        private AddBlockResult Suggest(Block? block, BlockHeader header, bool shouldProcess = true, bool? setAsMain = null, bool poSEnabled = false)
```
It is a temp hack. We planned to do something different with @marcindsobczak, but flags will be simpler and more elegant :)

paulhauner commented:
Hi there, I've been interop-ing with Lighthouse and discovered that the `engine_getPayload` response is missing the `extraData` field from [the spec](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.2/src/engine/interop/specification.md).

Below is the response from running the commands in this PR description.

```json
{
  "jsonrpc": "2.0",
  "result": {
    "gasLimit": "0x989680",
    "gasUsed": "0x0",
    "blockHash": "0xb084c10440f05f5a23a55d1d7ebcb1b3892935fb56f23cdc9a7f42c348eed174",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "random": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "coinbase": "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b",
    "blockNumber": "0x1",
    "parentHash": "0xa0513a503d5bd6e89a144c3268e5b7e9da9dbf63df125a360e3950a7d0d67131",
    "receiptRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
    "stateRoot": "0xca3149fa9e37db08d1cd49c9061db1002ef1cd58db2210f2115c8c989b2bdf45",
    "transactions": [],
    "timestamp": "0x5",
    "baseFeePerGas": "0x7"
  },
  "id": 67
}
```

The instructions to build Nethermind in the PR description were very helpful. Thank you :pray: It's my dumb mistake, but I forgot to run `git checkout themerge` and got a missing config error. Perhaps it might be helpful to add that to the commands.

I'm looking forward to interop-ing with Nethermind further!

MarekM25 commented:
> Hi there, I've been interop-ing with Lighthouse and discovered that the `engine_getPayload` response is missing the `extraData` field from [the spec](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.2/src/engine/interop/specification.md).
> 
> Below is the response from running the commands in this PR description.
> 
> ```json
> {
>   "jsonrpc": "2.0",
>   "result": {
>     "gasLimit": "0x989680",
>     "gasUsed": "0x0",
>     "blockHash": "0xb084c10440f05f5a23a55d1d7ebcb1b3892935fb56f23cdc9a7f42c348eed174",
>     "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
>     "random": "0x0000000000000000000000000000000000000000000000000000000000000000",
>     "coinbase": "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b",
>     "blockNumber": "0x1",
>     "parentHash": "0xa0513a503d5bd6e89a144c3268e5b7e9da9dbf63df125a360e3950a7d0d67131",
>     "receiptRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
>     "stateRoot": "0xca3149fa9e37db08d1cd49c9061db1002ef1cd58db2210f2115c8c989b2bdf45",
>     "transactions": [],
>     "timestamp": "0x5",
>     "baseFeePerGas": "0x7"
>   },
>   "id": 67
> }
> ```
> 
> The instructions to build Nethermind in the PR description were very helpful. Thank you 🙏 It's my dumb mistake, but I forgot to run `git checkout themerge` and got a missing config error. Perhaps it might be helpful to add that to the commands.
> 
> I'm looking forward to interop-ing with Nethermind further!

Thank you so much🙏 Sorry for forgetting about `git checkout themerge` I fixed the instruction and added the extra data field. It will be a pleasure for us to interop-ing with Lighthouse.

paulhauner commented:
Thanks @MarekM25, I can confirm the issue has been fixed! Lighthouse and Nethermind are now passing interop M2 :tada: 

You can find our docs here: https://github.com/sigp/lighthouse-merge-f2f/tree/master/m2_nethermind
